OrbitControls - Add new touch behavior (Google Maps like)

moroine opened the pull request:
Hello, 

This PR introduce a new touch behavior:

- 1 finger: pan
- 2 fingers:
    - spread or squish: zoom
    - rotate: orbit left
    - drag vertically: orbit up

The limitations of the implementation are:

- Starting orbit up will disable orbit left & zoom if order to prevent flickering
- The way to detect vertical drag might not be optimal and there is probably some constant fine tunning (I don't like that kind of stuff :stuck_out_tongue: ).

In order to keep the current one, I've also added a property `touchMode` which can take the following values:

- `THREE.TOUCH_MODE.CLASSIC`: the current one
- `THREE.TOUCH_MODE.MAP`: the one added, described above
- `THREE.TOUCH_MODE.LEGACY`: the one before #13706 i.e. 3 finger pan.

I have also changed the way the state is handled to use bit-mask and removed the STATE.TOUCH. I think using bit-mask could be useful to add something like the `mouseButtons` by customizing the bitmask used depending on the number of fingers like:

```javascript
touchFingers = {
  1: STATE.ROTATE,
  2: STATE.DOLLY | STATE.PAN
}
```

I have also updated the Orbit Control example: https://rawgit.com/moroine/three.js/OrbitControls-improvment/examples/misc_controls_orbit.html

Commit:
Add OrbitControls.touchMode to be able to use legacy behavior

Commit:
Fix regression

Commit:
Add Map Mode

Commit:
Merge remote-tracking branch 'upstream/dev' into OrbitControls-improvment

Commit:
Add details for each modes

Commit:
fix eslint

Commit:
Add example and fixes

Commit:
Add uglifyied build

Commit:
Fix rotateUp when drag two fingers in different ways

Previously when dragging fingers vertically in different directions (up / down) it was rotating up instead of rotating left

moroine commented:
@WestLangley what do you think about this ?

Gethi commented:
+1

stanley92 commented:
+1

WestLangley commented:
Live: https://rawgit.com/moroine/three.js/OrbitControls-improvment/examples/misc_controls_orbit.html

WestLangley commented:
I'm liking the support of "map mode". I'm not sure we need "legacy" mode, which uses the old three-finger panning.

Let's see what others think.

Commit:
Merge remote-tracking branch 'upstream/dev' into OrbitControls-improvment

Commit:
Fix linter issue after merge

moroine commented:
I've updated the current dev and removed the builds. 

About the legacy mode, I'm not sure too about "legacy" mode. Personally I don't need it, but I thought maybe some people want to use it in order to ensure backward compatibility on their projects.

WestLangley commented:
3-finger pan was removed in #13706. I do not think we need to reinsert it.

WestLangley commented:
>I've updated the current dev and removed the builds.

Really? There should only be 2 files modified in this PR.

moroine commented:
That's fine for me, and I am not a big fan on a mode called "legacy" :stuck_out_tongue:

Commit:
Remove legacy mode

Commit:
Restore build from mrdoob @dev

Commit:
Remove yarn.lock

moroine commented:
> Really? There should only be 2 files modified in this PR.

My bad I just removed the builds folder... Now should be ok, I've restored the one in dev branch and removed the yarn.lock included by mistake.

Do I remove the eslint fix for `src/renderers/webgl/WebGLAnimation.js` ?

moroine commented:
/ping @WestLangley

WestLangley commented on examples/js/controls/OrbitControls.js:
```diff
@@ -603,59 +683,157 @@ THREE.OrbitControls = function ( object, domElement ) {
+				//  - Vector rotateStart --> rotateEnd and rotateStart2 --> rotateEnd2 are in the same way (prevent moving one finger vertically up while the other goes down)
 
-			dollyStart.copy( dollyEnd );
+				var epsilonAngle = 1 / 2; // ~= sin(PI / 6)
```
This entire code block is too scary, IMO. Not sure what to do about it, though.

moroine replied:
Yeah, I'm not a big fan too

What about a refactor like this: 

```javascript
function handleTouchMoveRotate( event ) {

	//console.log( 'handleTouchMoveRotate' );

	if ( scope.enableRotate === false ) return;
	if ( ( state & STATE.ROTATE ) === 0 ) return;

	switch ( scope.touchMode ) {

		case THREE.TOUCH_MODES.CLASSIC:

			rotateEnd.set( event.touches[ 0 ].pageX, event.touches[ 0 ].pageY );

			rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );

			var element = scope.domElement === document ? scope.domElement.body : scope.domElement;

			rotateLeft( 2 * Math.PI * rotateDelta.x / element.clientHeight ); // yes, height

			rotateUp( 2 * Math.PI * rotateDelta.y / element.clientHeight );

			rotateStart.copy( rotateEnd );

			break;

		case THREE.TOUCH_MODES.MAP:

			// First finger
			var x = event.touches[ 0 ].pageX;
			var y = event.touches[ 0 ].pageY;

			rotateEnd.set( x, y );

			// Second finger
			var x2 = event.touches[ 1 ].pageX;
			var y2 = event.touches[ 1 ].pageY;

			rotateEnd2.set( x2, y2 );

			if ( isRotateUp() ) {

				var element = scope.domElement === document ? scope.domElement.body : scope.domElement;

				// rotating up and down along whole screen attempts to go 360, but limited to 180
				rotateUp( 2 * Math.PI * ( rotateEnd.y - rotateStart.y ) / element.clientHeight );

				// Start rotateUp ==> disable all movement to prevent flickering
				state = STATE.ROTATE_UP;

			} else if ( ( state & STATE.ROTATE_LEFT ) !== 0 ) {

				// Angle between the vector made by the two fingers and the positive x-axis at start
				var angleStartFingers = rotateDelta.subVectors( rotateStart2, rotateStart ).angle();
				// Angle between the vector made by the two fingers and the positive x-axis at end
				var angleEndFingers = rotateDelta.subVectors( rotateEnd2, rotateEnd ).angle();

				rotateLeft( ( angleStartFingers - angleEndFingers ) * scope.rotateSpeed );

			}

			rotateStart.copy( rotateEnd );
			rotateStart2.copy( rotateEnd2 );

			break;

		default:

			throw new Error( 'OrbitControls: Unexpected touchMode' );

	}

}

function isRotateUp() {

	// At start, does the two fingers are aligned horizontally
	if ( ! isHorizontal( rotateDelta.subVectors( rotateStart2, rotateStart ).angle(), Math.PI / 6 ) ) {

		return false;

	}

	// At end, does the two fingers are aligned horizontally
	if ( ! isHorizontal( rotateDelta.subVectors( rotateEnd2, rotateEnd ).angle(), Math.PI / 6 ) ) {

		return false;

	}

	// does the first finger moved vertically between start and end
	rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );
	if ( ! isVertical( rotateDelta.angle(), Math.PI / 6 ) ) {

		return false;

	}

	// does the second finger moved vertically between start and end
	rotateDelta2.subVectors( rotateEnd2, rotateStart2 ).multiplyScalar( scope.rotateSpeed );
	if ( ! isVertical( rotateDelta2.angle(), Math.PI / 6 ) ) {

		return false;

	}

	// Does the two finger moved in the same direction (prevent moving one finger vertically up while the other goes down)
	return rotateDelta.dot( rotateDelta2 ) > 0;

}

function isHorizontal( angleWithXAxis, precision ) {

	return Math.abs( Math.sin( angleWithXAxis ) ) < Math.sin( precision );

}

function isVertical( angleWithXAxis, precision ) {

	return Math.abs( Math.cos( angleWithXAxis ) ) < Math.cos( Math.PI / 2 - precision );

}
```

WestLangley commented:
>Do I remove the eslint fix for src/renderers/webgl/WebGLAnimation.js ?

It shouldn't really be there.

Commit:
Merge remote-tracking branch 'upstream/dev' into OrbitControls-improvment

moroine commented:
> It shouldn't really be there.

I merged dev branch and as it had been fixed, there is no more diff on that file.

WestLangley commented:
After talking to @mrdoob, it is preferred that we create `MapControls` separately. It will duplicate code, but that is OK... Or maybe `OrbitControls` can be extended?

Would you like to do a PR for that?

moroine commented:
Closed in favor of #14186
